MSMAll/scripts/SingleSubjectConcat.sh

```diff
@@ -31,504 +31,177 @@
-		error_count=$(( error_count + 1 ))
+log_Msg "Starting Single Subject Scan Concatenation using the selected frame range"
+# Naming Conventions and other variables
+IFS='@' read -a fMRINamesArray <<< "${fMRINames}"
+fMRINames=$(echo ${fMRINames} | sed 's/@/ /g')
+log_Msg "fMRINames: ${fMRINames}"
+
+if [ "${OutputProcSTRING}" = "NONE" ]; then
+	OutputProcSTRING=""
+fi
+log_Msg "OutputProcSTRING: ${OutputProcSTRING}"
+
+AtlasFolder="${StudyFolder}/${Subject}/MNINonLinear"
+log_Msg "AtlasFolder: ${AtlasFolder}"
+
+Caret7_Command=${CARET7DIR}/wb_command
+log_Msg "Caret7_Command: ${Caret7_Command}"
+
+# read frame informations
+runFrames=()
+runTR=()
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index))
+do
+	fMRIName="${fMRINamesArray[index]}"
+	ResultsFolder="${AtlasFolder}/Results/${fMRIName}"
+	log_Msg "ResultsFolder: ${ResultsFolder}"
+	NumFrames=`${Caret7_Command} -file-information "${ResultsFolder}/${fMRIName}${fMRIProcSTRING}.dtseries.nii" -only-number-of-maps`
+	runFrames[index]="$NumFrames"
+	TR=`${Caret7_Command} -file-information "${ResultsFolder}/${fMRIName}${fMRIProcSTRING}.dtseries.nii" -only-step-interval`
+	runTR[index]="$TR"
+	if [ $((index)) -eq 0 ]; then
+		minFrames=${runFrames[index]}
 	else
-		log_Msg "Variance Normalization: ${p_VarianceNormalization}"
+		(( ${runFrames[index]}<minFrames )) && minFrames=${runFrames[index]}
 	fi
-
-	if [ -z "${p_ComputeVarianceNormalization}" ]; then
-		log_Err "Compute Variance Normalization (--compute-variance-normalization=<YES | NO> required"
-		error_count=$(( error_count + 1 ))
+done
+if [[ "$EndFrame" == "" ]]; then
+	EndFrame="$minFrames"
+	log_Msg "EndFrame: ${EndFrame}"
+fi
+# check valid frame range
+if [ $((StartFrame)) -gt "${minFrames}" ] || [ $((EndFrame)) -gt "${minFrames}" ] || [ $((StartFrame)) -lt 1 ] || [ $((EndFrame)) -lt 1 ] || [ $((StartFrame)) -gt $((EndFrame)) ]; then
+	log_Err_Abort "The provided start frame ${StartFrame} and end frame ${EndFrame} is not valid, it must be between the range 1 to ${minFrames}. Please check the fMRI runs."
+fi
+log_Msg "StartFrame: ${StartFrame}, EndFrame: ${EndFrame}"
+FrameString="_Frame${StartFrame}To${EndFrame}"
+duration=0
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index))
+do
+	TR="${runTR[index]}"
+	Frames=$((EndFrame-StartFrame+1))
+	duration=$(bc -l <<< $duration+$TR*$Frames)
+done
+
+# convert to mins secs
+m=$(bc <<< "(${duration}%3600)/60")
+s=$((${duration%.*}%60))
+DurationString="_${m}mins${s}secs"
+
+cnt=0
+# loop for demean+vn+concat
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index)) ; do
+    fMRIName="${fMRINamesArray[$index]}"
+	log_Msg "fMRIName: ${fMRIName}"
+	ResultsFolder="${AtlasFolder}/Results/${fMRIName}"
+	log_Msg "ResultsFolder: ${ResultsFolder}"
+
+	# pick frames
+	if [ $((StartFrame)) -eq 1 ] && [ $((EndFrame)) -eq "${runFrames[index]}" ]; then
+		cp ${ResultsFolder}/${fMRIName}${fMRIProcSTRING}.dtseries.nii ${ResultsFolder}/${fMRIName}${fMRIProcSTRING}${FrameString}.dtseries.nii
```

Comment:
What is this line needed for?

Reply:
This condition means that the current frame range selects the full frame range of the data, which doesn't have to be put into -cifti-merge. This duplicate file is for consistency with the file to be processed afterwards, such as the file to set as temporary file, the file to use for demean and variance normalization. Otherwise, we may have duplicate operations for -cifti-merge, -cifti-reduce.

Reply:
Technically, doing a -cifti-merge in advance of the other code and concatenation will alter the provenance information.  However, it might simplify the logic to instead check whether the start and end frame raw options are something other than the default of "use everything", rather than trying to detect when the frame numbers specify the full run.

Reply:
However, making a copy of a file is a less efficient solution than using a variable to contain the file to use in later steps, and setting it appropriately according to whether a subset of the frames was requested.